Use QgsTextRenderer to render attribute table text

nyalldawson opened the pull request:
Allows for full range of text formatting in layout attribute tables, including buffers & shadows, word spacing, etc

Sponsored by City of Canning

nyalldawson committed:
[feature][layouts] Port attribute table to QgsTextRenderer

Allows for full range of text formatting in layout attribute tables,
including buffers & shadows, word spacing, etc

nyalldawson committed:
Ignore final line leading when calculating text height

nyalldawson committed:
Test masks

nyalldawson committed:
Scaled font also needs to scale letter/word spacing

nyalldawson committed:
Fix deprecation warnings and update unit tests

nirvn commented on src/core/layout/qgslayouttable.cpp:
```diff
@@ -612,31 +697,62 @@ void QgsLayoutTable::setShowEmptyRows( const bool showEmpty )
 
 void QgsLayoutTable::setHeaderFont( const QFont &font )
 {
-  if ( font == mHeaderFont )
```
@nyalldawson , any reason why you're not doing a font == mHeaderFont.toQFont() check here?

nyalldawson replied:
I'd also have to check the logic regarding size units too.. seems like overkill for a deprecated method which isn't used anywhere anymore!

nirvn replied:
@nyalldawson , fair enough.

nirvn commented on src/core/layout/qgslayouttable.cpp:
```diff
@@ -665,31 +781,62 @@ void QgsLayoutTable::setHeaderMode( const QgsLayoutTable::HeaderMode mode )
 
 void QgsLayoutTable::setContentFont( const QFont &font )
 {
-  if ( font == mContentFont )
```
See question above.

nirvn commented on src/core/layout/qgslayouttable.h:
```diff
@@ -628,40 +662,42 @@ class CORE_EXPORT QgsLayoutTable: public QgsLayoutMultiFrame
     /**
      * Draws the vertical grid lines for the table.
-     * \param painter destination painter for grid lines
+     * \param destination render context
```
```suggestion
     * \param context destination render context
```

nirvn commented on src/gui/layout/qgslayoutattributetablewidget.cpp:
```diff
@@ -72,8 +70,8 @@ QgsLayoutAttributeTableWidget::QgsLayoutAttributeTableWidget( QgsLayoutFrame *fr
 
-  mContentFontToolButton->setMode( QgsFontButton::ModeQFont );
-  mHeaderFontToolButton->setMode( QgsFontButton::ModeQFont );
+  mContentFontToolButton->setMode( QgsFontButton::ModeTextRenderer );
```
:tada:

nirvn commented on src/gui/qgsfontbutton.cpp:
```diff
@@ -869,6 +869,7 @@ void QgsFontButton::updatePreview( const QColor &color, QgsTextFormat *format, Q
       context.setUseAdvancedEffects( true );
       context.setFlag( QgsRenderContext::Antialiasing, true );
       context.setPainter( &p );
+      context.setFlag( QgsRenderContext::Antialiasing, true );
```
You set the antialiasing flag twice over three lines, should one setFlag() go away?

nyalldawson replied:
Ah, bad rebase sorry, fixed

nyalldawson committed:
Dox fix

nyalldawson commented:
Thanks @nirvn!

nyalldawson committed:
Test updates

nyalldawson committed:
Update vertical text with line spacing reference image

Note that the new image is more correct -- the previous one was incorrectly
rendering part of the label text outside of the registered label boundaries

nyalldawson committed:
Update masks

nyalldawson committed:
Test masks for Windows
